### ISSUE TITLE
EXT-X-BYTERANGE should come after other segment information

### ISSUE THREAD

**Issue by JacobLey** [kuu/hls-parser#98]
Hello this library has been a life-saver for me generating M3U8 files, but I've noticed they sometimes fail depending on the consumer, namely iOS/Safari based viewers.
Not that I believe this library's implementation that betrays the spec, but rather Apple is strict in ways they do not advertise (i.e. it is Apple who is wrong, but we can accommodate)

# The problem

Take two example M3U8 files:

(These files are slight modifications of sample files provided here: https://ottverse.com/free-hls-m3u8-test-urls/ -> http://demo.unified-streaming.com/video/tears-of-steel/tears-of-steel.ism/.m3u8)

`success.m3u8`
```
#EXTM3U
#EXT-X-VERSION:7
#EXT-X-TARGETDURATION:5
#EXT-X-PLAYLIST-TYPE:VOD
#EXTINF:4
http://demo.unified-streaming.com/video/tears-of-steel/tears-of-steel.ism/tears-of-steel-audio_eng=128002-video_eng=1001000-1.ts
#EXT-X-DISCONTINUITY
#EXTINF:4
#EXT-X-BYTERANGE:126712@0
http://demo.unified-streaming.com/video/tears-of-steel/tears-of-steel.ism/tears-of-steel-audio_eng=128002-video_eng=1001000-2.ts
#EXTINF:4
http://demo.unified-streaming.com/video/tears-of-steel/tears-of-steel.ism/tears-of-steel-audio_eng=128002-video_eng=1001000-3.ts
#EXT-X-ENDLIST
```

and `fail.m3u8`
```
#EXTM3U
#EXT-X-VERSION:7
#EXT-X-TARGETDURATION:5
#EXT-X-PLAYLIST-TYPE:VOD
#EXTINF:4
http://demo.unified-streaming.com/video/tears-of-steel/tears-of-steel.ism/tears-of-steel-audio_eng=128002-video_eng=1001000-1.ts
#EXT-X-BYTERANGE:126712@0
#EXT-X-DISCONTINUITY
#EXTINF:4
http://demo.unified-streaming.com/video/tears-of-steel/tears-of-steel.ism/tears-of-steel-audio_eng=128002-video_eng=1001000-2.ts
#EXTINF:4
http://demo.unified-streaming.com/video/tears-of-steel/tears-of-steel.ism/tears-of-steel-audio_eng=128002-video_eng=1001000-3.ts
#EXT-X-ENDLIST
```

Notice that:
1) The second segment has both a byte range (this is really the whole file, but proves a point here) and a discontinuity (again not actually necessary but proves a point).
2) The second file is _supposedly_ exactly the same, but the byte range is now at the start of the segment

Try loading the files in Safari (note Safari actually matters, I am not able to reproduce on Chrome, or other players like `hls.js` who are probably less strict)
A local serving package like https://www.npmjs.com/package/serve could come in handy

The first file is successful and plays a 12 second snippet, the second fails to load.

From my testing, it is the positioning of EXT-X-BYTERANGE before EXT-X-DISCONTINUITY (and potentially before EXTINF) that causes this issue

# How it impacts this package

The current stringify logic writes EXT-X-BYTERANGE before anything else https://github.com/kuu/hls-parser/blob/master/stringify.js#L296 recreating the issue discussed above.

As stated before I believe this is allowed by spec, but some interpreters do not handle it well, so it is easy enough to move that render to later in the `buildSegment` method.

I'm happy to submit a PR with this suggested change as well, but wanted to open the discussion for what the real problem is, and potential ways to fix it.

**Comment by kuu** [comment]
Thanks @JacobLey for letting me know this. I haven't noticed it.
I tested the playlists you provided with Safari and confirmed that `EXT-X-DISCONTINUITY` can't be placed after `EXT-X-BYTERANGE`.
As the playback works if I remove `EXT-X-DISCONTINUITY`, I thought putting `EXT-X-BYTERANGE` right after `EXT-X-DISCONTINUITY` would be enough.
But I agree that your PR (putting `EXT-X-BYTERANGE` right before the URI line) would be safer because of the following statements in the spec:
```
4.3.2.2.  EXT-X-BYTERANGE

...It applies only to the next URI line that follows it in the Playlist.
```

Thanks anyway. I'll merge your PR.

**Comment by kuu** [comment]
Published the change as hls-parser@0.10.2
